akka-projection-core/src/main/scala/akka/projection/HandlerRecovery.scala

```diff
@@ -33,56 +22,59 @@ object HandlerRecoveryStrategy {
 
   /**
    * If the first attempt to invoke the handler fails it will immediately give up,
    * discard the element and continue with next.
    */
-  def skip: HandlerRecoveryStrategy = Skip
+  def skip: HandlerRecoveryStrategy with AtMostOnceRecoveryStrategy with AtLeastOnceRecoveryStrategy = Skip
 
   /**
    * Scala API: If the first attempt to invoke the handler fails it will retry invoking the handler with the
    * same envelope this number of `retries` with the `delay` between each attempt. It will give up
    * and fail the stream if all attempts fail.
    */
-  def retryAndFail(retries: Int, delay: FiniteDuration): HandlerRecoveryStrategy =
+  def retryAndFail(retries: Int, delay: FiniteDuration): HandlerRecoveryStrategy with AtLeastOnceRecoveryStrategy =
     if (retries < 1) fail else RetryAndFail(retries, delay)
 
   /**
    * Java API: If the first attempt to invoke the handler fails it will retry invoking the handler with the
    * same envelope this number of `retries` with the `delay` between each attempt. It will give up
    * and fail the stream if all attempts fail.
    */
-  def retryAndFail(retries: Int, delay: java.time.Duration): HandlerRecoveryStrategy =
+  def retryAndFail(retries: Int, delay: java.time.Duration): HandlerRecoveryStrategy with AtLeastOnceRecoveryStrategy =
     retryAndFail(retries, delay.asScala)
 
   /**
    * Scala API: If the first attempt to invoke the handler fails it will retry invoking the handler with the
    * same envelope this number of `retries` with the `delay` between each attempt. It will give up,
    * discard the element and continue with next if all attempts fail.
    */
-  def retryAndSkip(retries: Int, delay: FiniteDuration): HandlerRecoveryStrategy =
+  def retryAndSkip(retries: Int, delay: FiniteDuration): HandlerRecoveryStrategy with AtLeastOnceRecoveryStrategy =
     if (retries < 1) fail else RetryAndSkip(retries, delay)
 
   /**
    * Java API: If the first attempt to invoke the handler fails it will retry invoking the handler with the
    * same envelope this number of `retries` with the `delay` between each attempt. It will give up,
    * discard the element and continue with next if all attempts fail.
    */
-  def retryAndSkip(retries: Int, delay: java.time.Duration): HandlerRecoveryStrategy =
+  def retryAndSkip(retries: Int, delay: java.time.Duration): HandlerRecoveryStrategy with AtLeastOnceRecoveryStrategy =
     retryAndSkip(retries, delay.asScala)
 
   /**
    * INTERNAL API: placed here instead of the `internal` package because of sealed trait
    */
   @InternalApi private[akka] object Internal {
-    case object Fail extends HandlerRecoveryStrategy
-    case object Skip extends HandlerRecoveryStrategy
-    final case class RetryAndFail(retries: Int, delay: FiniteDuration) extends HandlerRecoveryStrategy {
+    sealed trait AtMostOnceRecoveryStrategy extends HandlerRecoveryStrategy
+    sealed trait AtLeastOnceRecoveryStrategy extends HandlerRecoveryStrategy
```

Comment:
these can't be in Internal if they are exposed in public api

Reply:
:+1: